src/core/mount.c

```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
 
 DEFINE_STRING_TABLE_LOOKUP(mount_result, MountResult);
 
+static bool mount_can_freeze(Unit *u) {
+        Mount *m = MOUNT(u);
+        const char *fstype;
+
+        assert(u);
+        assert(m);
+
+        fstype = mount_get_fstype(m);
+        if (!fstype)
+                return false;
+
+        return fstype_can_freeze_thaw(fstype);
+}
+
+static int mount_freezer_action(Unit *u, FreezerAction action) {
+        _cleanup_close_ int fd = -EBADF;
+        Mount *m = MOUNT(u);
+        FreezerState next, target;
+        int r;
+
+        assert(u);
+        assert(m);
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_PARENT_FREEZE,
+                              FREEZER_THAW, FREEZER_PARENT_THAW));
+
+        unit_next_freezer_state(u, action, &next, &target);
+        next = freezer_state_finish(next); /* We're completely sync */
```

Comment:
the ioctl is indeed sync, like all disk accesses. we could theoroetically do this async, i.e. fork off a thread or so. but maybe it's not worth the effort for now, given that disk accesses are pretty much always sync, dunno.

Reply:
As far as I can tell, filesystem freezing should be instant; I don't think it recurses down the tree like a cgroup freeze does. But of course the implementation is filesystem dependent, so :shrug: 

I think it makes most sense to leave it sync for now, and then if someone discovers a filesystem that does actually recursively do things in response to a freeze request then we make it async

CC @brauner